### ISSUE TITLE
Array() is not shared and causes various issues

### ISSUE THREAD

**Issue by kirilledelman** [godotengine/godot#4114]
I found a bug with arrays / dictionaries. Consider this code example:

```
var a = []
a.push_back( { dictionaryValue = 2 } )
print ( "a.size: ", a.size() )
# prints a.size: 1

var b = Array()
b.push_back( { dictionaryValue = 2 } )
print ( "b.size: ", b.size() )
# prints b.size: 1

var x = { items = [] }
x.items.push_back( { dictionaryValue = 2 } )
print ( "x.items.size: ", x.items.size() )
# prints x.size: 1

var y = { items = Array() }
y.items.push_back( { dictionaryValue = 2 } )
print ( "y.items.size: ", y.items.size() )
# prints y.size: 0
```

y.items.size should also be 1

**Comment by bojidar-bg** [comment]
First of all, I can confirm.

Here is something weirder:
````
var y = { items = Array() }
y.items.push_back( { dictionaryValue = 2 } )
print ( "y.items.size: ", y.items.size() )
# prints y.size: 0

var z = { items = Array() }
var z_a = z.items
z_a.push_back( { dictionaryValue = 2 } )

print ( "z.items.size: ", y.items.size() )
# prints z.size: 0
print ( "z_a.size: ", z_a.size() )
#!! prints z_a.size: 1
````

Also, it isn't related to the array being contained by the Dictionary class either, since
````
var w = [Array()]
w[0].push_back({ dictionaryValue = 2 })
print(w[0].size())
# 0
````

Neither it is related to the inner value being a Dictionary, since:
````
var w = [Array()]
w[0].push_back(1)
print(w[0].size())
# 0
````

**Comment by toger5** [comment]
This is already discussed in the in issue #2015
Reduz:

> just get used to use [] instead of Array(), or you will constantly face whit issue

it seems the problem is when you are using `Array()` to create a new array. as both of you are doing. If you use 
```
var w = [[]]
w[0].push_back(1)
print("and it will print: ",w[0].size())
#and it will print: 1
```
I just ried it out and it works.
Maybe `Array()` creates a specific array which doesn't allow to store datatypes like other arrays?

**Comment by bojidar-bg** [comment]
@toger5 ~ From the same issue it says that `Array()`-created arrays are not shared, unlike `array`.

@akien-mga ~ I'm not really sure which one of the two should be marked as a dupe. (also, add 2.1 to the other one)

**Comment by kirilledelman** [comment]
Now I get the same bug elsewhere when trying to remove items. This is a serious issue. 
My array that can't be modified comes from unserializing a string via str2var, and is inside a dictionary. Can't add or remove items from this array. No errors. Please fix.

**Comment by toger5** [comment]
@reduz :
> This bug is because by default Array() is not shared, this will change eventually in 2.1. However, using [] instead of array should work

**Comment by kirilledelman** [comment]
@toger5 I don't think it does
![fucksakes](https://cloud.githubusercontent.com/assets/6539713/14054000/89443e6c-f2a3-11e5-83dc-9e493cb40fcc.jpg)

**Comment by reduz** [comment]
This is tagged to be fixed in Godot 3.0 as it means breaking some compatibility to fix

**Comment by reduz** [comment]
As promised, this has been fixed for a while in 3.0, closing
